### PR TITLE
[WIP DO NOT MERGE] update to work with Python 3.10.

### PR DESCRIPTION
**Pull request by ethanluoyc**

### CODE REVIEW

**Comment by ethanluoyc** [comment]
@kenjitoyama I just updated the PR and have tested this to compile for Python 3.10. This should be good enough for a start. 

The main issues around updating the dependencies should be sorted, but there are some things which are just quite hard to do it with hacks on my side. Specifically, I am not entirely sure what's the approach to work with python packaging on top of bazel right now, so the PR currently has some code that is a bit coupled with py310. Let me unpack. I added some comments to the diff so perhaps you can help me take a look?

File: .gitignore
Comment (ethanluoyc) [inline .gitignore:1]: Added for gitignore sanity.

File: docker/Dockerfile
Comment (ethanluoyc) [inline docker/Dockerfile:1]: Only updated for the single Dockerfile, maybe we can discuss what base to use for the allversions Dockerfile as well.

File: docker/Dockerfile [hunk 1]
```diff
@@ -7,18 +7,18 @@ RUN apt update
...
+RUN chmod +x /bazel-6.1.0-linux-x86_64
+RUN mv /bazel-6.1.0-linux-x86_64 /usr/bin/bazel
 
 # Add python alternatives.
-RUN update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.8 1
+RUN update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.10 1
 
 # Override gcc/g++.
-RUN update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-8 60 --slave /usr/bin/g++ g++ /usr/bin/g++-8
+RUN update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-9 60 --slave /usr/bin/g++ g++ /usr/bin/g++-9
```
Comment (ethanluoyc) [inline docker/Dockerfile:21]: gcc-9 as per Tensorflow 2.12.0

https://www.tensorflow.org/install/source

File: envlogger/BUILD.bazel [hunk 2]
```diff
@@ -15,9 +15,18 @@
 # Build targets for environment logger.
 load("@py_deps//:requirements.bzl", "requirement")
 load("@rules_python//python:defs.bzl", "py_library")
+load("@rules_python//python:pip.bzl", "compile_pip_requirements")
 
 package(default_visibility = ["//visibility:public"])
 
+# This rule adds a convenient way to update the requirements file.
+compile_pip_requirements(
```
Comment (ethanluoyc) [inline envlogger/BUILD.bazel:23]: If we use `requirement`, then it expects a compiled dependency (generated below).

This is a hack! I couldn't find a good way to create requirements locks with multiple Python versions.

Comment (ethanluoyc) [reply envlogger/BUILD.bazel:23]: I think the rules_python pip repository does not work well with repositories that have native extensions

File: docker/Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM ubuntu:18.04
+FROM ubuntu:22.04
```
Comment (kenjitoyama) [inline docker/Dockerfile:1]: TF uses `20.04`. If we use a version that's too recent, its glibc may not work on older versions of TF. Does this work fine with `20.04`?

**Comment by ethanluoyc** [comment]
I got rid of all requirements as it was easier to first install these into the build environment in OSS. This is what launchpad and reverb do as well. I am not sure how much the BUILD.bazel shares with blaze (I guess not much) but this might be something you would like to fix internally. In launchpad and reverb I think there is some build_def.bzl sharing with blaze.

The cross_language_test fails when running with bazel test because for some reason the way it passed the runfile environment variables removes PATH information from the subprocess call which causes the py_writer to resolve to the python under /usr/bin/python. When running outside the container, this causes the tests to fail.

I think the code mostly works so we can start thinking about cleaning it up?

**Comment by kenjitoyama** [comment]
Hi @ethanluoyc , sorry for the delay I was OOO.

What's the current state now? Does this build the 3.10 wheels successfully?

Thanks,

Daniel

**Comment by ethanluoyc** [comment]
> Hi @ethanluoyc , sorry for the delay I was OOO.
> 
> What's the current state now? Does this build the 3.10 wheels successfully?
> 
> Thanks,
> 
> Daniel

Hi! 
Yes, it should work.

**Comment by kenjitoyama** [comment]
This work has been absorbed into https://github.com/deepmind/envlogger/commit/1c0d8f29c114128e75573ea7a605eeb0548ca29b.

**Pull request closed without merging**